### PR TITLE
only add coord mask if mask is set in raster

### LINKED ISSUE #350: BUG: Memory error when loading large raster datasets clipped to a geometry
### HydroMT version checks

- [X] I have checked that this issue has not already been reported.
- [X] I have checked that this bug exists on the latest version of HydroMT.   
- [X] I have checked that this bug exists on the main branch of HydroMT.            

### Issue Description

When loading a rasterdataset clipped with a geom we get memory errors for large datasets because of the `geometry_mask` in `raster.clip_geom` method, see [raster.py#L1157](https://github.com/Deltares/hydromt/blob/e4edd37a19f89bd99614782272864bad58527b27/hydromt/raster.py#L1157) With HydroMT-SFINCS we use many highres datasets which currently are impacted by this bug.

TODO
- [ ] Analyze what impact it has further downstream to remove the geometry_mask in `raster.clip_geom`.
- [ ] Use `clip_bbox` instead of `clip_geom` in the `RasterDatasetAdapter.get_data` method. The bbox can be calculated as `bbox = geom.to_crs(4326).total_bounds`.

### Reproducible Example

```
from hydromt import DataCatalog 
from shapely.geometry import box
import geopandas as gpd

bbox = [-90, 10, -70, 50]
gdf = gpd.GeoDataFrame(geometry=[box(*bbox)], crs=4326) 
cat = DataCatalog(r"deltares_data")
da0 = cat.get_rasterdataset("esa_worldcover", bbox=bbox) # no error
da1 = cat.get_rasterdataset("esa_worldcover", geom=gdf) # memory error
```

### Installed Versions

hydromt 0.7.1

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by savente93**
## Issue addressed
Fixes #350 

## Explanation
When trying to clip a rasterized dataset to a geometry, the `clip_geom` had to allocate an entire numpy array at the highest resolution everywhere leading to memory issues when using high res datasets. To solve this problem, we moved the called to `geometry_mask ` inside the `clip_geom` to only happen if a mask is actually set. This solves the problem, but might have consequences for hydromt_wflow, I will open an issue there for them to investigate if this has impact for them. 

## Checklist
- [x] Updated tests or added new tests
- [x] Branch is up to date with `main`
- [x] Tests & pre-commit hooks pass
- [x] Updated documentation if needed

## Additional Notes (optional)
Add any additional notes or information that may be helpful.

### CODE REVIEW

**Review by DirkEilander — APPROVED** [review]
The OR looks good to me! Can you just update the changelog?
I suggest to wait for a response on https://github.com/Deltares/hydromt_wflow/issues/187 before merging.

**Comment by savente93** [comment]
closing in favour of #434 as that fixes more instances of this issue

**Pull request closed without merging**
